<launch>
    <arg name="sim" default="false"/>
    <arg name="use_ros_gmapping" default="false"/>

    <param name="/use_sim_time" value="true"/>
    <node name="dream_gmapping_odometer" type="dream_gmapping_odometer" pkg="dream_gmapping" output="screen">
        <param name="wheel_pos_topic" value="/dream/wheel_pos"/>
        <param name="wheel_dist" value="0.156"/>
        <param name="wheel_diameter" value="0.041"/>
        <param name="base_frame" value="base_link"/>
        <param name="odom_frame" value="odom"/>
    </node>

<group unless="$(arg use_ros_gmapping)">
    <node name="dream_gmapper" type="dream_gmapper" pkg="dream_gmapping" output="screen" >
        <param name="base_frame" value="base_link"/>
        <param name="map_frame" value="map"/>
        <param name="map_update_interval" value="5"/>
        <param name="particle_num" value="4"/>
        <param name="translation_active_threshold" value="0.01" if="$(arg sim)"/>
        <param name="angular_active_threshold" value="0.1"/>

        <param name="d_v_std_dev" value="0.0001"/>
        <param name="d_theta_std_dev" value="0.0001"/>
        <param name="resolution" value="0.05"/>
        <param name="beam_noise_variance" value="0.04"/>
        <param name="half_beam_kernel_size" value="2"/>
        <param name="skip_invalid_beams" value="true"/>
        <param name="pose_correction_kernel_size" value="3"/>
        <!-- min value for hit/total_counts to be considered as an obstacle -->
        <param name="occupied_fullness_threshold" value="0.6"/>


        <!-- params for sim -->
        <param name="max_range" value="6.0" if="$(arg sim)"/>
        <param name="scan_topic" value="/dream/scan"/>
        <param name="map_topic" value="/map"/>
        <param name="map_size_in_meters" value="20"/>
        <param name="publish_debug_scan" value="false"/>
    </node>
</group>

<group if="$(arg use_ros_gmapping)">
    <!-- http://docs.ros.org/en/hydro/api/gmapping/html/index.html -->
    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen" >
      <remap from="scan" to="/dream/scan"/>
      <param name="map_update_interval" value="5.0"/>
      <param name="maxUrange" value="6.0"/>
      <param name="sigma" value="0.05"/>
      <param name="kernelSize" value="1"/>
      <param name="lstep" value="0.05"/>
      <param name="astep" value="0.05"/>
      <param name="iterations" value="5"/>
      <param name="lsigma" value="0.075"/>
      <param name="ogain" value="3.0"/>
      <param name="lskip" value="0"/>
      <param name="srr" value="0.1"/>
      <param name="srt" value="0.2"/>
      <param name="str" value="0.1"/>
      <param name="stt" value="0.2"/>
      <param name="linearUpdate" value="0.1"/>
      <param name="angularUpdate" value="0.1"/>
      <param name="temporalUpdate" value="3.0"/>
      <param name="resampleThreshold" value="0.5"/>
      <param name="particles" value="30"/>
      <param name="xmin" value="-10.0"/>
      <param name="ymin" value="-10.0"/>
      <param name="xmax" value="10.0"/>
      <param name="ymax" value="10.0"/>
      <param name="delta" value="0.05"/>
      <param name="llsamplerange" value="0.01"/>
      <param name="llsamplestep" value="0.01"/>
      <param name="lasamplerange" value="0.005"/>
      <param name="lasamplestep" value="0.005"/>
    </node>
</group>
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find dream_gmapping)/rviz/view_dream_gmapper.rviz"/>
</launch>
